src/umpire/strategy/ZeroByteHandler.cpp

```diff
@@ -0,0 +1,88 @@
+//////////////////////////////////////////////////////////////////////////////
+// Copyright (c) 2016-19, Lawrence Livermore National Security, LLC and Umpire
+// project contributors. See the COPYRIGHT file for details.
+//
+// SPDX-License-Identifier: (MIT)
+//////////////////////////////////////////////////////////////////////////////
+#include "umpire/strategy/ZeroByteHandler.hpp"
+
+#include "umpire/util/Macros.hpp"
+
+#include "umpire/ResourceManager.hpp"
+
+#include "umpire/strategy/FixedPool.hpp"
+
+namespace umpire {
+namespace strategy {
+
+ZeroByteHandler::ZeroByteHandler(
+  std::unique_ptr<AllocationStrategy>&& allocator) noexcept :
+AllocationStrategy(allocator->getName(), allocator->getId()),
+m_allocator(std::move(allocator))
+{
+}
+
+void*
+ZeroByteHandler::allocate(std::size_t bytes)
+{
+  if (0 == bytes) {
+    UMPIRE_LOG(Debug, "Allocating 0 bytes for" << m_allocator->getName());
+    auto& rm = ResourceManager::getInstance();
+    return rm.getZeroByteAllocator()->allocate(1);
+  } else {
+    return m_allocator->allocate(bytes);
+  }
+}
+
+void
+ZeroByteHandler::deallocate(void* ptr)
+{
+  auto& rm = ResourceManager::getInstance();
+  auto zero_pool = dynamic_cast<FixedPool*>(rm.getZeroByteAllocator());
+
+  if (zero_pool->pointerIsFromPool(ptr)) {
+    UMPIRE_LOG(Debug, "Deallocating 0 bytes for" << m_allocator->getName());
+    zero_pool->deallocate(ptr);
+  } else {
+    m_allocator->deallocate(ptr);
+  }
+}
```

Comment:
This seems backwards to me.  I would suggest checking the zero pool in the catch block of `m_allocator->deallocate()`.  That way, there is no additional cost for normal deallocations.

Reply:
That adds a try-catch though, right?

Reply:
Correct.  What is the problem with that?  The catch would be expensive, but only for deallocations of 0 byte pointers.  Checking for 0 byte pointers for every deallocate impacts most of the allocations.